### PR TITLE
release-23.1: colexec: fix incorrect eager cancellation in parallel unordered sync

### PR DESCRIPTION
**Pull request by yuzefovich**
Backport 1/1 commits from #127076.

/cc @cockroachdb/release

---

As of 2a928d0d0d8cbe5e2f86224b404bdbbbcfb23f49 we have eager cancellation of work in the parallel unordered sync for local plans. This is an optimization to cancel outstanding work in some inputs whenever the sync is being drained as a whole (and is needed to power the locality-optimized search feature). The cancellation is achieved by cancelling the contexts of all inputs and then ignoring all errors from those inputs.

However, if we have multiple parallel unordered syncs in the plan, then eager cancellation in the ancestor can "poison" the execution of the descendant because the latter cannot distinguish between benign errors (that happened because of the eager cancellation) and true query errors. (The sync that performs the eager cancellation distinguishes between the two by checking whether its "root" context is canceled or not.) Other operators don't have this issue since they all run within a single goroutine, with the exception of the lookup and index joins that use the streamer, but there the main "read" goroutine is the same one as for other operators that use a single goroutine, so the parallel sync can correctly filter out benign errors.

This commit fixes an issue of spurious context cancellation errors during valid query execution by allowing eager cancellation on drain only in "leaf" parallel syncs. This should be sufficient for the locality-optimized search feature where we have multiple table readers at the same level combined by the sync (and table readers themselves are always leaf operators).

Fixes: #127043.

Release note (bug fix): CockroachDB previously could encounter spurious `(error encountered after some results were delivered)` `ERROR: context canceled` errors when evaluating some queries in rare cases. Most likely conditions are:
- the query has to be executed locally
- have a LIMIT
- have at least two UNION clauses
- have some lookup / index joins in the UNION branches.

The bug is present since 22.2 and is now fixed.

Release justification: bug fix.

### CODE REVIEW

**Review by michae2 — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127658#-:-O2kGo2yDfvFlrsC1QMr:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

**Comment by yuzefovich** [comment]
@mgartner as we agreed, not merging without your approval - please merge this when you take a look

**Review by rafiss — CHANGES_REQUESTED** [review]
if this gets merged, please also include this test change to make the test less flaky: https://github.com/cockroachdb/cockroach/pull/131680

**Comment by yuzefovich** [comment]
will be backporting together with the fix for #127942

**Pull request closed without merging**
